### ISSUE TITLE
No warning returning optional parameter of unconstrained type

### ISSUE THREAD

**Issue by cston** [dotnet/roslyn#40818]
No warning is reported for the second method:
```C#
#nullable enable

class Program
{
    static T F<T>()
    {
        T defaultValue = default;
        return defaultValue; // warning CS8603: Possible null reference return
    }

    static T F<T>(T defaultValue = default)
    {
        return defaultValue; // no warning
    }
}
```
From https://github.com/dotnet/roslyn/issues/38339#issuecomment-569571346, updated with warnings from 16.5. 
cc @Bosch-Eli-Black

**Comment by Bosch-Eli-Black** [comment]
Thanks for filing this, @cston :)

**Comment by jcouv** [comment]
It feels like the warning should be on the optional parameter value: `static T F<T>(T defaultValue = default) // here`. No?

**Comment by jaredpar** [comment]
Verified that this repros even after our defaultable generics work.

**Comment by RikkiGibson** [comment]
> It feels like the warning should be on the optional parameter value: static T F<T>(T defaultValue = default) // here. No?

We do not warn for `T t = default`, so I think the solution here needs to be to initialize the parameter state based on the default parameter value. We had discussed doing this in the past but put it off, assuming that we would warn on the default parameter value if it was bad going in.
